File: 678.有效的括号字符串.py
#
# @lc app=leetcode.cn id=678 lang=python3
#
# [678] 有效的括号字符串
#

# @lc code=start
class Solution:
    def checkValidString(self, s: str) -> bool:
        left, aster = [], []
        length = len(s)
        for i in range(length):
            if s[i] == "(":
                left.append(i)
            elif s[i] == "*":
                aster.append(i)
            elif len(left) > 0: # 右括号匹配一个左括号
                left = left[:len(left)-1]
            elif len(aster) > 0:    # 无左括号则匹配一个 *
                aster = aster[:len(aster)-1]
            else:   # 都无则 False
                return False
        i = len(left) - 1
        j = len(aster) - 1
        while i >= 0 and j >= 0:
            if left[i] > aster[j]:
                return False
            i, j = i - 1, j - 1
        # 小于 0 则完整匹配结束
        return i < 0
# @lc code=end

File: 678.有效的括号字符串.go
/*
 * @lc app=leetcode.cn id=678 lang=golang
 *
 * [678] 有效的括号字符串
 */

// @lc code=start
func checkValidString(s string) bool {
	var left, aster []int
	length := len(s)
	for i:=0;i<length;i++ {
		if s[i] == '(' {
			left = append(left, i)
		} else if s[i] == '*' {
			aster = append(aster, i)
		} else if len(left) > 0 {
			left = left[:len(left)-1]
		} else if len(aster) > 0 {
			aster = aster[:len(aster)-1]
		} else {
			return false
		}
	}
	i := len(left) - 1
	for j := len(aster) - 1;i>=0&&j>=0;i,j=i-1,j-1 {
		if left[i] > aster[j] {
			return false
		}
	}
	return i < 0
}
// @lc code=end

